libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,564 @@
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		yulAssert(_op1.location == Location::Memory, "");
+		if (_op1.length && knowledge.knownToBeZero(*_op1.length))
+			return true;
+		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
+			return true;
+		if (!_op1.start || !_op2.start || !_op1.length || !_op2.length)
+			// Can we say anything else here?
+			return false;
+
+		// 1.start + 1.length <= 2.start ||
+		if (optional<u256> length1 = knowledge.knownConstant(*_op1.length))
+			if (optional<u256> diff = knowledge.differenceIsKnownConstant(*_op2.start, *_op1.start))
+				// diff = 2.start - 1.start
+				if (*length1 <= *diff && *length1 <= largestPositive && *diff <= largestPositive)
+				{
+					//cout << "Knon unrelated: length1 " << *length1 << ", diff: " << *diff << endl;
+					return true;
+				}
+		// 2.start + 2.length <= 1.start
+		if (optional<u256> length2 = knowledge.knownConstant(*_op2.length))
+			if (optional<u256> diff = knowledge.differenceIsKnownConstant(*_op1.start, *_op2.start))
+				// diff = 1.start - 2.start
+				if (*length2 <= *diff && *length2 <= largestPositive && *diff <= largestPositive)
+				{
+					//cout << "Knon unrelated: length2 " << *length2 << ", diff: " << *diff << endl;
+					return true;
+				}
+	}
+
+	return false;
+}
+
+bool RedundantStoreEliminator::knownCovered(
+	RedundantStoreEliminator::Operation const& _covered,
+	RedundantStoreEliminator::Operation const& _covering
+) const
+{
+	if (_covered.location != _covering.location)
+		return false;
+	if (_covered.start && _covered.length && _covered.start == _covering.start && _covered.length == _covering.length)
+		return true;
+	if (_covered.location == Location::Storage && _covered.start && _covering.start && *_covered.start == *_covering.start)
+		return true;
+	if (_covered.location == Location::Memory)
+	{
+		KnowledgeBase knowledge(m_dialect, m_ssaValues);
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		// i.start <= e.start && e.start + e.length <= i.start + i.length
+		if (!_covered.start || !_covering.start || !_covered.length || !_covering.length)
+			// Can we say anything else here?
+			return false;
+		// TODO i.start is zero, this can still be done, even if the difference is not a constant.
+		// TODO this does not work well when "sub" is involved - do we need more simplification rules?
+		if (optional<u256> startDiff = knowledge.differenceIsKnownConstant(*_covered.start, *_covering.start))
+			if (optional<u256> lengthDiff = knowledge.differenceIsKnownConstant(*_covering.length, *_covered.length))
+				if (*startDiff <= *lengthDiff && *startDiff <= largestPositive && *lengthDiff <= largestPositive)
+					return true;
```

Comment:
Maybe you can add a test that would lead to an incorrect removal and mention EIP-1985?

```
{
    let x := 0

    calldatacopy(0, 0, 115792089237316195423570985008687907853269984665640564039457584007913129639935)
    mstore(x, 20)
    return(0, 32)
}
// ----
// step: redundantStoreEliminator
//
// {
//     let x := 0
//     let _1 := 115792089237316195423570985008687907853269984665640564039457584007913129639935
//     let _2 := 0
//     let _3 := 0
//     mstore(x, 20)
//     return(0, 32)
// }
```

Reply:
I think it can have bugs like these when writing to locations beyond `2**128 - 1`. This is a grey area because of EIP-1985.